tools/write_changelog_entry.py

```diff
@@ -0,0 +1,68 @@
+import os
+import re
+import glob
+
+
+def inc(obj, key):
+    """Increment a dictionary[key] value."""
+    try:
+        obj[key] += 1
+    except KeyError:
+        obj[key] = 1
+
+
+# Load PR number and comment
+try:
+    pr = os.environ['PR_NUMBER']
+    body = os.environ['PR_BODY']
+except KeyError:
+    raise ValueError('PR_NUMBER and PR_BODY env vars must be set.')
+assert str(int(pr)) == pr
+body = body.replace('\\r', '').replace('\\n', '\n')
+
+# Get changelog directory
+# TODO: get directory from pyproject.toml
+path = os.path.join(os.getenv('CHANGELOG_DIR', 'changelog'))
+if not os.path.exists(path):
+    raise ValueError(f"CHANGELOG_DIR path '{path}' must exist.")
+
+# Search for changelog entries
+#     markdown heading of "Changelog [category]"
+#     followed by a message in the next ``` code block.
+body = re.sub("(<!--.*?-->)", "", body, flags=re.DOTALL)
+regex = r"[#]+[ ]*Changelog[ ]*\[([a-zA-Z]+)\]*[^`]*```([^`]*)```"
+results = re.findall(regex, body)
+
+# Count entries per category
+count = {}
+for category, _ in results:
+    inc(count, category)
+
+# Generates (filename, message) pairs
+entries = []  # list of changelogs
+n = {}  # counter for each category
+for category, text in results:
+
+    if count[category] > 1:  # include filename [.COUNTER]
+        inc(n, category)
+        counter = f'.{n[category]}'
+    else:
+        counter = ''
+
+    filename = f"{pr}.{category}{counter}.rst"
+    text = text.strip()
+    if len(text) > 0:  # ignore empty messages
+        entries += [(filename, text)]
+
+for existing in glob.glob(os.path.join(path, f"{pr}.*.rst")):
```

Comment:
I am trying to think of a way we can detect that the action added the file rather than just nuking all changelogs.

Could we use git to pull the commit message of the last commit editing the file and see if we have a unique string in that message? That means that if the last commit wasn't the changelog bot then we don't delete it and instead we error?

Reply:
Could run e.g. `git log -s -n1 --pretty='format:%an %ae%n' changelog/6032.feature.rst` on each file before deleting, and only delete if it's the SunPyBot?

Reply:
Actually, it would be much simpler (and with less surprises) if this workflow can only commit if *all* the existing changelog entries for the PR were last edited by the bot. (Because otherwise it's complicated to handle multiple changelogs of the same type where some are controlled by the bot and some are manually controlled.)

Reply:
That would be fine with me as well.

Reply:
Which would you prefer? Bot command or top comment? Top comment would be easiest to implement (as that's what we have already), although happy to switch to bot command if you would prefer.

Reply:
I can't find a reliable way to check if the comment author should have permission to push to the PR branch (as the workflow will always have permission). So, if you don't mind, I'll just keep it as updating the top comment so permissions can be enforced. I'll add the check mentioned above.